File: python/src/ancilis/deps/manifest.py
"""Manifest detection and dependency parsing for common Python manifest formats."""

from __future__ import annotations

import json
import re
import sys
from dataclasses import dataclass, field
from pathlib import Path

if sys.version_info >= (3, 11):
    import tomllib
else:
    try:
        import tomllib  # type: ignore[import]
    except ImportError:
        tomllib = None  # type: ignore[assignment]


@dataclass
class Dependency:
    name: str
    version: str | None
    source_file: str


@dataclass
class Manifest:
    path: str
    format: str
    dependencies: list[Dependency] = field(default_factory=list)


# Matches pinned lines like: requests==2.28.0, requests==2.28.0[security]
_REQUIREMENTS_PINNED = re.compile(
    r"^\s*([A-Za-z0-9_.\-]+)(?:\[.*?\])?==([^\s;#]+)", re.IGNORECASE
)


def _parse_requirements_txt(path: Path) -> list[Dependency]:
    deps: list[Dependency] = []
    for line in path.read_text(encoding="utf-8").splitlines():
        line = line.strip()
        if not line or line.startswith("#") or line.startswith("-r") or line.startswith("-c"):
            continue
        m = _REQUIREMENTS_PINNED.match(line)
        if m:
            deps.append(Dependency(name=m.group(1), version=m.group(2), source_file=str(path)))
    return deps


def _normalise_pep508(spec: str) -> tuple[str, str | None]:
    """Extract (name, version) from a PEP 508 specifier, or (name, None) if unpinned."""
    # Strip extras, environment markers
    spec = re.sub(r"\[.*?\]", "", spec).split(";")[0].strip()
    m = re.match(r"^([A-Za-z0-9_.\-]+)==([^\s,]+)", spec)
    if m:
        return m.group(1), m.group(2)
    m = re.match(r"^([A-Za-z0-9_.\-]+)", spec)
    if m:
        return m.group(1), None
    return spec, None


def _parse_pyproject_toml(path: Path) -> list[Dependency]:
    if tomllib is None:
        return []
    try:
        data = tomllib.loads(path.read_text(encoding="utf-8"))
    except Exception:
        return []
    raw_deps = data.get("project", {}).get("dependencies", [])
    deps: list[Dependency] = []
    for spec in raw_deps:
        name, version = _normalise_pep508(spec)
        if version is not None:
            deps.append(Dependency(name=name, version=version, source_file=str(path)))
    return deps


def _parse_pipfile_lock(path: Path) -> list[Dependency]:
    try:
        data = json.loads(path.read_text(encoding="utf-8"))
    except (json.JSONDecodeError, OSError):
        return []
    deps: list[Dependency] = []
    for section in ("default", "develop"):
        for pkg, info in data.get(section, {}).items():
            if isinstance(info, dict):
                version_raw = info.get("version", "")
                # Pipfile.lock versions look like "==2.28.0"
                m = re.match(r"==(.+)", version_raw)
                if m:
                    deps.append(Dependency(name=pkg, version=m.group(1), source_file=str(path)))
    return deps


def _parse_poetry_lock(path: Path) -> list[Dependency]:
    if tomllib is None:
        return []
    try:
        data = tomllib.loads(path.read_text(encoding="utf-8"))
    except Exception:
        return []
    deps: list[Dependency] = []
    for pkg in data.get("package", []):
        name = pkg.get("name")
        version = pkg.get("version")
        if name and version:
            deps.append(Dependency(name=name, version=str(version), source_file=str(path)))
    return deps


class ManifestDetector:
    """Detect and parse Python dependency manifests in a project directory."""

    _HANDLERS: list[tuple[str, str, object]] = [
        ("requirements.txt", "requirements.txt", _parse_requirements_txt),
        ("pyproject.toml", "pyproject.toml", _parse_pyproject_toml),
        ("Pipfile.lock", "Pipfile.lock", _parse_pipfile_lock),
        ("poetry.lock", "poetry.lock", _parse_poetry_lock),
    ]

    def detect(self, project_dir: Path) -> list[Manifest]:
        """Find and parse all supported manifest files in *project_dir*."""
        manifests: list[Manifest] = []
        for filename, fmt, handler in self._HANDLERS:
            candidate = project_dir / filename
            if candidate.is_file():
                deps = handler(candidate)  # type: ignore[operator]
                manifests.append(Manifest(path=str(candidate), format=fmt, dependencies=deps))
        return manifests

File: typescript/src/ancilis/deps/manifest.ts
/** Manifest detection and dependency parsing for common Node.js manifest formats. */

import { readFileSync, existsSync } from "node:fs";
import { join } from "node:path";
import { parse as parseYaml } from "yaml";
import type { Dependency, Manifest } from "./types.js";

// ----- package-lock.json -----

function parsePackageLockJson(path: string): Dependency[] {
  let data: Record<string, unknown>;
  try {
    data = JSON.parse(readFileSync(path, "utf-8")) as Record<string, unknown>;
  } catch {
    return [];
  }

  const deps: Dependency[] = [];
  const lockVersion = (data["lockfileVersion"] as number | undefined) ?? 1;

  if (lockVersion >= 2 && data["packages"] && typeof data["packages"] === "object") {
    // v2/v3: packages is { "node_modules/name": { version: "x.y.z" } }
    for (const [key, info] of Object.entries(data["packages"] as Record<string, unknown>)) {
      if (!key) continue; // skip root entry ""
      if (!(info && typeof info === "object")) continue;
      const pkg = info as Record<string, unknown>;
      const version = pkg["version"] as string | undefined;
      if (!version) continue;
      // Strip "node_modules/" prefix, possibly nested like "node_modules/a/node_modules/b"
      const name = key.replace(/^.*node_modules\//, "");
      deps.push({ name, version, sourceFile: path });
    }
  } else if (data["dependencies"] && typeof data["dependencies"] === "object") {
    // v1: dependencies is { "name": { version: "x.y.z" } }
    for (const [name, info] of Object.entries(data["dependencies"] as Record<string, unknown>)) {
      if (!(info && typeof info === "object")) continue;
      const pkg = info as Record<string, unknown>;
      const version = pkg["version"] as string | undefined;
      if (version) {
        deps.push({ name, version, sourceFile: path });
      }
    }
  }

  return deps;
}

// ----- yarn.lock -----

function parseYarnLock(path: string): Dependency[] {
  let content: string;
  try {
    content = readFileSync(path, "utf-8");
  } catch {
    return [];
  }

  const deps: Dependency[] = [];
  const seen = new Set<string>();
  const lines = content.split("\n");

  let currentName: string | null = null;

  for (const line of lines) {
    // Skip comment and blank lines
    if (line.startsWith("#") || line.trim() === "") {
      currentName = null;
      continue;
    }

    // Package specifier: not indented, ends with ":"
    if (!line.startsWith(" ") && !line.startsWith("\t") && line.endsWith(":")) {
      // May be: "pkg@^1.0.0": or pkg@^1.0.0: or "@scope/pkg@^1.0.0":
      const raw = line.slice(0, -1).replace(/^"|"$/g, "").trim();
      // Take the first specifier if comma-separated
      const firstSpec = raw.split(",")[0]!.trim().replace(/^"|"$/g, "").trim();
      // The name is everything before the last "@" (handles scoped packages "@scope/name@version")
      const lastAt = firstSpec.lastIndexOf("@");
      currentName = lastAt > 0 ? firstSpec.slice(0, lastAt) : null;
      continue;
    }

    // Version line inside a block: "  version "x.y.z""
    if (currentName !== null) {
      const vMatch = /^\s+version\s+"([^"]+)"/.exec(line);
      if (vMatch && !seen.has(currentName)) {
        seen.add(currentName);
        deps.push({ name: currentName, version: vMatch[1]!, sourceFile: path });
        currentName = null;
      }
    }
  }

  return deps;
}

// ----- pnpm-lock.yaml -----

function parsePnpmLock(path: string): Dependency[] {
  let content: string;
  try {
    content = readFileSync(path, "utf-8");
  } catch {
    return [];
  }

  let data: Record<string, unknown>;
  try {
    data = parseYaml(content) as Record<string, unknown>;
  } catch {
    return [];
  }

  // v5–v6 use "packages"; v9 uses "snapshots" for fully resolved packages
  const section = (data["packages"] ?? data["snapshots"]) as Record<string, unknown> | undefined;
  if (!section || typeof section !== "object") return [];

  const deps: Dependency[] = [];

  for (const key of Object.keys(section)) {
    // v5: "/lodash@4.17.21", v6+: "lodash@4.17.21", possibly with qualifiers "(patch_hash)"
    const cleanKey = key.replace(/^\//, "");
    // Match "name@version" — name may be scoped "@scope/pkg"
    const match = /^(@?[^@(]+)@([^@()\s]+)/.exec(cleanKey);
    if (match) {
      const name = match[1]!;
      const version = match[2]!;
      deps.push({ name, version, sourceFile: path });
    }
  }

  return deps;
}

// ----- package.json (fallback) -----

const PINNED_VERSION = /^\d/;

function parsePackageJson(path: string): Dependency[] {
  let data: Record<string, unknown>;
  try {
    data = JSON.parse(readFileSync(path, "utf-8")) as Record<string, unknown>;
  } catch {
    return [];
  }

  const deps: Dependency[] = [];

  for (const section of ["dependencies", "devDependencies"] as const) {
    const entries = data[section] as Record<string, string> | undefined;
    if (!entries || typeof entries !== "object") continue;
    for (const [name, version] of Object.entries(entries)) {
      if (typeof version === "string" && PINNED_VERSION.test(version)) {
        deps.push({ name, version, sourceFile: path });
      }
    }
  }

  return deps;
}

// ----- ManifestDetector -----

type Handler = (path: string) => Dependency[];

const HANDLERS: Array<[filename: string, format: string, handler: Handler]> = [
  ["package-lock.json", "package-lock.json", parsePackageLockJson],
  ["yarn.lock", "yarn.lock", parseYarnLock],
  ["pnpm-lock.yaml", "pnpm-lock.yaml", parsePnpmLock],
  ["package.json", "package.json", parsePackageJson],
];

export class ManifestDetector {
  /** Find and parse all supported manifest files in `projectDir`. */
  detect(projectDir: string): Manifest[] {
    const manifests: Manifest[] = [];

    for (const [filename, format, handler] of HANDLERS) {
      const candidate = join(projectDir, filename);
      if (existsSync(candidate)) {
        const dependencies = handler(candidate);
        manifests.push({ path: candidate, format, dependencies });
      }
    }

    return manifests;
  }
}
